### PR TITLE
Mark contiguity of reduction ID as nullopt

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Reduction IDs does not exist, so it makes no sense to mark them as true or false.

Also:
- Add prove for `i1 % size < size`
- Add simplification `i1 / d < D` -> `i1 < d * D`

The above two items is important to make the predicate of reduction unrelated to contiguity.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
@naoyam Do you think my changes here make sense? I don't know what a contiguity for a reduction means. But this PR is **NOT** a no-op. I do see changes like `i1 < size1 * size2` -> `i1 / size2 < size1 && i1 % size2 < size2`. (BTW, the part after `&&` is trivial, if this PR's change does make sense, I will update this PR to make expr simplifier simplify it away.)

**Comment by zasdfgbnm** [comment]
Maybe instead of doing `value_or(false)`, I should do `value_or(true)`? I can not think of a case where merged reductions's indices can not be collapsed.

**Comment by zasdfgbnm** [comment]
Yeah, `i1 / size2 < size1` is mathematically equivalent to `i1 < size1 * size2`, so I think there are two separate things:
1. Expr simplifier should be capable of simplifying the former to the latter, because `*` is faster than `/`, with this simplification, I don't really care if the contiguity of a reduction should be considered true or false.
2. Still, the contiguity of a reduction should be stored as `std::nullopt` but treated as `true` during indexing to be double safe.

**Comment by naoyam** [comment]
I think it makes sense to have `nullopt` or something for reduction domains. Maybe we should also remove them from allocation domains?

**Comment by zasdfgbnm** [comment]
> > If I remove reductions from allocation domain, indexing won't be happy.
> 
> Do you know why? No need to dig into, just curious if you already know.

Some of the `contig_finder.allocToIndexedID().find(root_id)` like thing will fail. I don't remember where.

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
@naoyam ready for review

File: csrc/expr_simplifier.cpp [hunk 1]
```diff
@@ -94,23 +94,28 @@ class Logger : public NoOpLogger {
       : NoOpLogger(value), init_val_(value), current_val_(value) {}
 
   ~Logger() override {
-    if (!shouldPrint()) {
-      return;
-    }
+    try {
```
Comment (zasdfgbnm) [inline csrc/expr_simplifier.cpp:97]: Suggest reading hiding whitespace

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**